[MIPS][float] Fixed SingleFloat codegen on N32/N64 targets

Tazdevil971 opened the pull request:
This patch aims at making the combination of single-float and N32/N64 ABI properly work.

Right now when both options are enabled the compiler chooses an incorrect ABI and in some cases even generates wrong instructions.

The floating point behavior on MIPS is controlled through 3 flags: soft-float, single-float, fp64. This makes things complicated because fp64 indicates the presence of 64bit floating point registers, but cannot be easily disabled (the mips3 feature require it, but mips3 CPUs with only 32bit floating point exist). Also if fp64 is missing it doesn't actually disable 64bit floating point operations, because certain MIPS1/2 CPUs support 64bit floating point with 32bit registers, hence the single-float option.

I'm guessing that originally single-float was only intended for the latter case, and that's the reason why it doesn't properly work on 64bit targets.

So this patch does the following:
- Make single-float a "master disable", even if fp64 is enabled this should completely disable generation of 64bit floating point operations, making it available on targets which hard require fp64.
- Add proper calling conventions for N32/N64 single-float combinations.
- Fixup codegen to not generate certain 64bit floating point operations, apparently not assigning a register class to f64 values is not enough to prevent them from showing up.
- Add tests for the new calling conventions and codegen.

Tazdevil971 committed:
[MIPS][float] Fixed SingleFloat codegen on N32/N64 targets

arsenm commented on llvm/test/CodeGen/Mips/cconv/arguments-hard-float-single-varargs.ll:
```diff
@@ -0,0 +1,148 @@
+define void @double_args(double %a, ...)
+                         nounwind {
+entry:
+        %0 = getelementptr [11 x double], ptr @doubles, i32 0, i32 1
+        store volatile double %a, ptr %0
+
+        %ap = alloca ptr
+        call void @llvm.va_start(ptr %ap)
+        %b = va_arg ptr %ap, double
+        %1 = getelementptr [11 x double], ptr @doubles, i32 0, i32 2
+        store volatile double %b, ptr %1
+        call void @llvm.va_end(ptr %ap)
+        ret void
```
Weird / too much indentation, and should use named values in tests

Tazdevil971 replied:
This is because this tests are updated versions of the soft-float/hard-float variants.

If you look at how tests are currently structured, for example `arguments-float.ll`, there are two versions `arguments-float.ll` which tests soft-float behavior, then `arguments-hard-float.ll` for hard-float behavior, but the underlying test is identical.

So what I did was add a third case for hard single-float, and updated the outcome. So if they look weird it's probably because the test I took this code from use outdated coding conventions.

I didn't want to update the old tests, so I think the best option is to just leave all of this as is.

The only thing I could change is naming, because now that I look at them, all the -hard tests put -hard right after the first word, while I put -single at the end of the path.

So to be more consistent I could change the name to `arguments-hard-single-float-varargs.ll` for example.

arsenm replied:
So the IR content is all the same? You could cleanup the old tests and switch them to generated tests, and add new run lines? I guess that will still be a very long test with a few too many run combinations

Tazdevil971 replied:
Yes the IR is the same. 

This test need to already deal with 3 ABIs (O32/N32/N64) and 2 endianess, plus if we put all of the combinations of floating point flags this would cram 18 different run combinations in a single test file which I think is wayyyy to many.

But I could switch all of them to generated tests, unless there is a specific reason I'm missing for why they were hand-written. As you can see some registers are captured and matched multiple times, instead of being hardcoded, maybe this is to make the test agnostic to register allocation?

Also I don't quite understand what you mean by "named values in tests" in your first comment, sorry I'm not familiar with test code conventions.

Tazdevil971 replied:
I tried merging all of the different test cases, honestly it's not that bad, you can see it [here](https://github.com/llvm/llvm-project/blob/8fda8677250b1bb1d3fc476deb8329c55c8b3942/llvm/test/CodeGen/Mips/cconv/arguments-float.ll). You can actually ignore BE/LE variants and end up with just 9 different run combinations.

If you are okay with it I could merge all of the test variants and do some refactoring of the old tests touched by this PR.

arsenm commented on llvm/lib/Target/Mips/MipsISelLowering.cpp:
```diff
@@ -4295,10 +4295,16 @@ parseRegForInlineAsmConstraint(StringRef C, MVT VT) const {
+    // otherwise if the size of FP registers is 64-bit or Reg is an even number,
+    // select the 64-bit register class. Otherwise, select the 32-bit register
+    // class.
+    if (VT == MVT::Other) {
+      if (Subtarget.isSingleFloat())
+        VT = MVT::f32;
+      else
+        VT = (Subtarget.isFP64bit() || !(Reg % 2)) ? MVT::f64 : MVT::f32;
```
Not sure this is properly covered in the tests, the only asm is a full clobber list which I'm assuming is to stress calling convention handling

Tazdevil971 replied:
You are right, I'll prepare a test case for it

brad0 commented:
cc @yingopq @wzssyqa

Tazdevil971 commented:
Ping @arsenm.

Recap:
- I've added missing test cases and renamed existing ones
- I haven't touched the existing tests as that is still an open question, [here](https://github.com/llvm/llvm-project/blob/8fda8677250b1bb1d3fc476deb8329c55c8b3942/llvm/test/CodeGen/Mips/cconv/arguments-float.ll) I tried doing it. If you think it looks better that way I can do it on all of them and add it to the PR.

brad0 commented:
@yingopq

brad0 commented:
@arsenm @yingopq Ping.

Tazdevil971 commented on llvm/test/CodeGen/Mips/cconv/arguments-hard-single-float-varargs.ll:
```diff
@@ -0,0 +1,148 @@
+define void @double_args(double %a, ...)
+                         nounwind {
+entry:
+        %0 = getelementptr [11 x double], ptr @doubles, i32 0, i32 1
```
Same as [https://github.com/llvm/llvm-project/pull/140575#discussion_r2382315392](here)

Tazdevil971 commented on llvm/lib/Target/Mips/MipsSEISelLowering.cpp:
```diff
@@ -211,6 +212,20 @@ MipsSETargetLowering::MipsSETargetLowering(const MipsTargetMachine &TM,
+    setOperationAction(ISD::FP_TO_SINT, MVT::i64, Expand);
+    setOperationAction(ISD::FP_TO_UINT, MVT::i64, Expand);
+    setOperationAction(ISD::STRICT_FP_TO_SINT, MVT::i64, Expand);
+    setOperationAction(ISD::STRICT_FP_TO_UINT, MVT::i64, Expand);
```
Yes you are right, I checked and strictfp doesn't seem to be supported by MIPS so I just removed it

wzssyqa reviewed (commented):
N32 uses the same FP ABI like N64.

Tazdevil971 commented:
I rechecked online and it doesn't appear to be the case. First let me preface this by saying that the N32 has been mostly been lost to time, and I could find very few things on the internet.

That said, online I found this: [MIPSproTM N32 ABI handbook](https://irix7.com/techpubs/007-2816-004.pdf), which is what I used to figure out what parts of the ABI where missing.

From that book, at page 14, it highlights some differences between the 3 ABIs (O32/N32 and N64, called the 64-bit interface). There you can see that callee saved registers behave weirdly in N32, only even registers from f20 to f31 are callee saved.

Also you can see the same thing at page 10, where it highlights caller/callee saved registers.

Finally it also says the same thing at page 9:
> There are eight callee-saved floating point registers, $f24..$f31 for the 64-bit interface. There are six for the n32 ABI, the six even registers in $f20..$f30.
> [The o32-bit ABI specifies the six even registers, or even/odd pairs, $f20..$f31.]

And if we look at what gcc does, it also uses the same "call used regs" for N32 and O32 FP64, and you can see it [here](https://github.com/gcc-mirror/gcc/blob/e1b9ccaa10df0192d8c4818258510821727a3de2/gcc/config/mips/mips.cc#L21123) (I hope this is the right place, I'm not very familiar with the gcc codebase).

So no, in my opinion, N32 and N64 do not use the same FP ABI. Instead the N32 FP ABI is a lot more like the O32 F64 ABI.

Tazdevil971 commented on llvm/lib/Target/Mips/MipsCallingConv.td:
```diff
@@ -357,13 +359,19 @@ def CSR_O32_FP64 :
+                  RA_64, FP_64, GP_64, (sequence "S%u_64", 7, 0))>;
+
+def CSR_N32_SingleFloat
+    : CalleeSavedRegs<(add(decimate(sequence "F%u", 30, 20), 2), RA_64, FP_64,
```
I wrote a summary of why I don't think this is the case [here](https://github.com/llvm/llvm-project/pull/140575#issuecomment-3342058261).
